python/pyspark/sql/dataframe.py

```diff
@@ -1761,12 +1761,37 @@ def toPandas(self):
                 raise ImportError("%s\n%s" % (e.message, msg))
         else:
             dtype = {}
+            columns_with_null_int = set()
+            def null_handler(rows, columns_with_null_int):
+                for row in rows:
+                    row = row.asDict()
+                    for column in columns_with_null_int:
+                        val = row[column]
+                        dt = dtype[column]
+                        if val is not None:
+                            if abs(val) > 16777216: # Max value before np.float32 loses precision.
+                                val = np.float64(val)
+                                dt = np.float64
+                                dtype[column] = np.float64
+                            else:
+                                val = np.float32(val)
+                                if dt not in (np.float32, np.float64):
+                                    dt = np.float32
+                                    dtype[column] = np.float32
+                            row[column] = val
+                    row = Row(**row)
+                    yield row
+            row_handler = lambda x,y: x
             for field in self.schema:
                 pandas_type = _to_corrected_pandas_type(field.dataType)
+                if pandas_type in (np.int8, np.int16, np.int32) and field.nullable:
+                    columns_with_null_int.add(field.name)
+                    row_handler = null_handler
+                    pandas_type = np.float32
```

Comment:
I don't think this is a correct fix.

Reply:
Can you elaborate? I believe it is, per my reply to your comment in the `null_handler`.

Reply:
Have you read carefully the comments in https://github.com/apache/spark/pull/18945#discussion_r134033952, https://github.com/apache/spark/pull/18945#discussion_r134925269? They are good suggestions for this issue. I don't know why you don't want to follow them to check null values with Pandas...

Reply:
A simple wrong for this line is, even this condition is met, don't necessarily meaning there are null values in the column. But you forcibly set the type to np.float32.

Reply:
Ah, I see where I got confused. I had started with @ueshin 's suggestion but abandoned it because I didn't want to create the DataFrame before the type correction because I was also looking at @HyukjinKwon 's suggestion. I somehow ended up combining them incorrectly.

Reply:
I will take my suggestion back. I think thier suggestions are better than mine.